algorithms/src/std_algorithms/Kokkos_MinMaxElementOperations.hpp

```diff
@@ -0,0 +1,420 @@
+  }
+
+  // aliases
+  using index_type   = typename IteratorType::difference_type;
+  using value_type   = typename IteratorType::value_type;
+  using reducer_type = typename std::conditional<
+      (0 < sizeof...(Args)),
+      ReducerType<value_type, index_type, Args..., ExecutionSpace>,
+      ReducerType<value_type, index_type, ExecutionSpace> >::type;
+
+  using result_view_type = typename reducer_type::result_view_type;
+  using func_t           = StdMinMaxElemFunctor<IteratorType, reducer_type>;
+
+  // run
+  result_view_type result("minmax_elem_impl_result");
+  reducer_type reducer(result, std::forward<Args>(args)...);
+  const auto num_elements = last - first;
+  ::Kokkos::parallel_reduce(label,
+                            RangePolicy<ExecutionSpace>(ex, 0, num_elements),
+                            func_t(first, reducer), reducer);
+  ex.fence("minmax_element: fence after operation");
+
+  // return
+  const auto result_h =
+      ::Kokkos::create_mirror_view_and_copy(::Kokkos::HostSpace(), result);
+  return {first + result_h().min_loc, first + result_h().max_loc};
+}
+
+}  // end namespace Impl
+
+// ----------------------
+// min_element public API
+// ----------------------
+template <class ExecutionSpace, class IteratorType>
+auto min_element(const ExecutionSpace& ex, IteratorType first,
+                 IteratorType last) {
+  return Impl::min_or_max_element_impl<MinFirstLoc>(
+      "kokkos_min_element_iterator_api_default", ex, first, last);
+}
+
+template <class ExecutionSpace, class IteratorType>
+auto min_element(const std::string& label, const ExecutionSpace& ex,
+                 IteratorType first, IteratorType last) {
+  return Impl::min_or_max_element_impl<MinFirstLoc>(label, ex, first, last);
+}
+
+template <class ExecutionSpace, class IteratorType, class ComparatorType>
+auto min_element(const ExecutionSpace& ex, IteratorType first,
+                 IteratorType last, ComparatorType comp) {
+  static_assert_is_not_opemnptarget(ex);
+
+  return Impl::min_or_max_element_impl<MinFirstLocCustomComparator>(
+      "kokkos_min_element_iterator_api_default", ex, first, last,
+      ::Kokkos::Experimental::move(comp));
+}
+
+template <class ExecutionSpace, class IteratorType, class ComparatorType>
+auto min_element(const std::string& label, const ExecutionSpace& ex,
+                 IteratorType first, IteratorType last, ComparatorType comp) {
+  static_assert_is_not_opemnptarget(ex);
+
+  return Impl::min_or_max_element_impl<MinFirstLocCustomComparator>(
+      label, ex, first, last, ::Kokkos::Experimental::move(comp));
+}
+
+template <class ExecutionSpace, class DataType, class... Properties>
+auto min_element(const ExecutionSpace& ex,
+                 const ::Kokkos::View<DataType, Properties...>& v) {
+  static_assert_is_admissible_to_kokkos_std_algorithms(v);
+
+  return Impl::min_or_max_element_impl<MinFirstLoc>(
+      "kokkos_min_element_view_api_default", ex, cbegin(v), cend(v));
```

Comment:
I don't think that is right to pass a constant iterator.  I think the constness should derive from the view data type.

Reply:
thanks for the comment. I actually thing that it helps, or does not hurt in any way. 
Nothing changes to the user, they can pass a view with read only or read-write, but it helps us behind the scenes because it "forces us" to implement min/max such that we don't violate constness. Min/max is intrinsically something that requires const, so I don't see why we should allow something that is not const.

Reply:
Except that in that in that case it influences the return type of the function

Reply:
True, we will change it in cases where this affects the return type and try to keep it where it's harmless.

Reply:
oh right!